### PR TITLE
Arm64: Add TransposeVectorEven and TransposeVectorOdd intrinsics

### PR DESCRIPTION
**Pull request by TamarChristinaArm**
Hi All,

This adds `TRN1` and `TRN2`.
As with #1190 these intrinsics haven't gone through API review yet and there is
some overlap with #1190 so whichever gets merged first I'll rebase the other one.

I am wondering about the name though, In the proposal I called `TransposeVectorEven`/`TransposeVectorOdd`,
but now I'm wondering if it should be `TransposeEven` and `TransposeOdd` since that's more consistent with
Zip* and Unzip*.

Implements https://github.com/dotnet/corefx/issues/42176

/cc @CarolEidt @tannergooding @echesakovMSFT

Thanks,
Tamar

### CODE REVIEW

**Comment by BruceForstall** [comment]
@TamarChristinaArm fyi, looks like you have a conflict to resolve.

**Comment by TamarChristinaArm** [comment]
Thanks @BruceForstall , that should take care of it.

**Comment by echesakovMSFT** [comment]
Oops - I just merged another Tamar's PR (#1190) and this caused another set of conflicts. Sorry about that. 
I guess we gonna have this type of situation for a while since we are touching the same files while working on intrinsics.

**Comment by echesakovMSFT** [comment]
> I am wondering about the name though, In the proposal I called TransposeVectorEven/TransposeVectorOdd,
but now I'm wondering if it should be TransposeEven and TransposeOdd since that's more consistent with
Zip* and Unzip*.

@TamarChristinaArm I personally like `TransposeEven` `TransposeOdd` names better.

File: src/coreclr/src/jit/instrsarm64.h [hunk 1]
```diff
@@ -1221,6 +1221,12 @@ INST1(umaxv,   "umaxv",  0, 0, IF_DV_2M,  0x2E30A800)
 INST1(uminv,   "uminv",  0, 0, IF_DV_2M,  0x2E31A800)
                                    //  uminv   Vd,Vn                DV_2M  0Q101110XX110001 101010nnnnnddddd   2E31 A800   Vd,Vn      (vector)
 
+INST1(trn1,    "trn1",   0, 0, IF_DV_3A,  0x0E002800)
+                                   //  zip1    Vd,Vn,Vm             DV_3A  0Q001110XX0mmmmm 001010nnnnnddddd   0E00 2800   Vd,Vn,Vm  (vector)
```
Comment (echesakovMSFT) [inline src/coreclr/src/jit/instrsarm64.h]: Shouldn't the comment be about `trn1` and not `zip1` instruction?

File: src/coreclr/src/jit/instrsarm64.h [hunk 2]
```diff
@@ -1221,6 +1221,12 @@ INST1(umaxv,   "umaxv",  0, 0, IF_DV_2M,  0x2E30A800)
...
+
+INST1(trn2,    "trn2",   0, 0, IF_DV_3A,  0x0E006800)
+                                   //  zip2    Vd,Vn,Vm             DV_3A  0Q001110XX0mmmmm 011010nnnnnddddd   0E00 6800   Vd,Vn,Vm  (vector)
```
Comment (echesakovMSFT) [inline src/coreclr/src/jit/instrsarm64.h]: Same here - `trn2` and not `zip2` instruction?

File: src/coreclr/src/jit/emitarm64.cpp [hunk 3]
```diff
@@ -12892,7 +12896,17 @@ emitter::insExecutionCharacteristics emitter::getInsExecutionCharacteristics(ins
                 case INS_umax:
                 case INS_smin:
                 case INS_smax:
-                    result.insThroughput = PERFSCORE_THROUGHPUT_2X;
+                case INS_trn1:
+                case INS_trn2:
+                    if ((ins == INS_trn1 || ins == INS_trn1)
+                        && (id->idInsOpt() == INS_OPTS_2D))
+                    {
+                        result.insThroughput = PERFSCORE_THROUGHPUT_1C;
```
Comment (echesakovMSFT) [inline src/coreclr/src/jit/emitarm64.cpp]: Out of curiosity - where are you getting these numbers from?

Comment (TamarChristinaArm) [reply src/coreclr/src/jit/emitarm64.cpp]: @briansull added the PerfScore using Cortex-A55 as a base, so they can be found https://developer.arm.com/docs/epm128372/30/arm-cortex-a55-software-optimization-guide here.

File: src/coreclr/src/jit/instrsarm64.h (hunk 1 above)
Comment (TamarChristinaArm) [reply src/coreclr/src/jit/instrsarm64.h]: Yes sorry, had somehow missed updating that..

**Comment by TamarChristinaArm** [comment]
> @TamarChristinaArm I personally like TransposeEven TransposeOdd names better.

Cool, If there aren't any objections I will update it tomorrow then.

> Oops - I just merged another Tamar's PR (#1190) and this caused another set of conflicts. Sorry about that.

No worries, I was expecting that conflict :)

**Review by briansull — APPROVED** [review]
LGTM

**Comment by TamarChristinaArm** [comment]
I've renamed the intrinsics to `TransposeEven` and `TransposeOdd`. Is it ok for me to edit https://github.com/dotnet/corefx/issues/42176 still or should I close and re-open a new one in runtime?

**Comment by tannergooding** [comment]
It should be fine to edit the existing API proposal since it isn't marked `api-approved`; if it had been, 
we would want to open a new issue (or maybe a centralized tracking issue) instead.

**Comment by TamarChristinaArm** [comment]
> It should be fine to edit the existing API proposal since it isn't marked api-approved

Done, proposal has been updated with these new names.

File: src/coreclr/src/jit/emitarm64.cpp [hunk 4]
```diff
@@ -12906,6 +12908,20 @@ emitter::insExecutionCharacteristics emitter::getInsExecutionCharacteristics(ins
                     result.insLatency    = PERFSCORE_LATENCY_2C;
                     break;
 
+                case INS_trn1:
+                case INS_trn2:
+                    if ((ins == INS_trn1 || ins == INS_trn1) && (id->idInsOpt() == INS_OPTS_2D))
```
Comment (tannergooding) [inline src/coreclr/src/jit/emitarm64.cpp]: you are doing the same comparison twice, its also not clear why the comparison here is needed given the above cases.

**Review by tannergooding — APPROVED** [review]
LGTM, modulo the comment abotu double comparison

File: src/coreclr/src/jit/emitarm64.cpp (hunk 4 above)
Comment (Mistuke) [reply src/coreclr/src/jit/emitarm64.cpp]: ah, the rebasing did have a casualty.. I'll fix that up.

Comment (TamarChristinaArm) [reply src/coreclr/src/jit/emitarm64.cpp]: ah, the rebasing did have a casualty.. I'll fix that up. The if was because there was a fall through before.

**Review by echesakov — APPROVED** [review]
LGTM

**Pull request merged**